File: flyweight/java/Flyweight.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// Flyweight: Character formatting
class CharacterFormat {
    private String font;
    private int size;
    private String color;

    public CharacterFormat(String font, int size, String color) {
        this.font = font;
        this.size = size;
        this.color = color;
    }

    public String display(char character) {
        return "Character '" + character + "' [Font: " + font + ", Size: " + size + ", Color: " + color + "]";
    }
}

// Flyweight Factory
class FormatFactory {
    private Map<String, CharacterFormat> formats;

    public FormatFactory() {
        this.formats = new HashMap<>();
    }

    public CharacterFormat getFormat(String font, int size, String color) {
        String key = font + "-" + size + "-" + color;
        if (!formats.containsKey(key)) {
            System.out.println("Created new format: " + key);
            formats.put(key, new CharacterFormat(font, size, color));
        }
        return formats.get(key);
    }

    public int getTotalFormats() {
        return formats.size();
    }
}

// Character with extrinsic state
class Character {
    private char character;
    private CharacterFormat format;

    public Character(char character, CharacterFormat format) {
        this.character = character;
        this.format = format;
    }

    public String display() {
        return format.display(character);
    }
}

// Text Editor
class TextEditor {
    private List<Character> characters;
    private FormatFactory factory;

    public TextEditor() {
        this.characters = new ArrayList<>();
        this.factory = new FormatFactory();
    }

    public void addCharacter(char character, String font, int size, String color) {
        CharacterFormat format = factory.getFormat(font, size, color);
        characters.add(new Character(character, format));
    }

    public List<String> display() {
        List<String> results = new ArrayList<>();
        for (Character character : characters) {
            results.add(character.display());
        }
        results.add("Total unique formats: " + factory.getTotalFormats());
        return results;
    }
}

// Main class demonstrating Flyweight pattern
public class Flyweight {
    public static void main(String[] args) {
        TextEditor editor = new TextEditor();
        editor.addCharacter('H', "Arial", 12, "Red");
        editor.addCharacter('e', "Arial", 12, "Red");
        editor.addCharacter('l', "Arial", 12, "Red");
        editor.addCharacter('l', "Times", 14, "Blue");
        editor.addCharacter('o', "Times", 14, "Blue");

        List<String> results = editor.display();
        for (String result : results) {
            System.out.println(result);
        }
    }
}

File: flyweight/typescript/Flyweight.ts
// Flyweight: Character formatting
class CharacterFormat {
  private font: string;
  private size: number;
  private color: string;

  constructor(font: string, size: number, color: string) {
    this.font = font;
    this.size = size;
    this.color = color;
  }

  display(char: string): string {
    return `Character '${char}' [Font: ${this.font}, Size: ${this.size}, Color: ${this.color}]`;
  }
}

// Flyweight Factory
class FormatFactory {
  private formats: Map<string, CharacterFormat>;

  constructor() {
    this.formats = new Map();
  }

  getFormat(font: string, size: number, color: string): CharacterFormat {
    const key = `${font}-${size}-${color}`;
    if (!this.formats.has(key)) {
      console.log(`Created new format: ${key}`);
      this.formats.set(key, new CharacterFormat(font, size, color));
    }
    return this.formats.get(key)!;
  }

  getTotalFormats(): number {
    return this.formats.size;
  }
}

// Character with extrinsic state
class Character {
  private char: string;
  private format: CharacterFormat;

  constructor(char: string, format: CharacterFormat) {
    this.char = char;
    this.format = format;
  }

  display(): string {
    return this.format.display(this.char);
  }
}

// Text Editor
class TextEditor {
  private characters: Character[];
  private factory: FormatFactory;

  constructor() {
    this.characters = [];
    this.factory = new FormatFactory();
  }

  addCharacter(char: string, font: string, size: number, color: string): void {
    const format = this.factory.getFormat(font, size, color);
    this.characters.push(new Character(char, format));
  }

  display(): string[] {
    const results = this.characters.map(char => char.display());
    results.push(`Total unique formats: ${this.factory.getTotalFormats()}`);
    return results;
  }
}

// Usage
const editor = new TextEditor();
editor.addCharacter('H', 'Arial', 12, 'Red');
editor.addCharacter('e', 'Arial', 12, 'Red');
editor.addCharacter('l', 'Arial', 12, 'Red');
editor.addCharacter('l', 'Times', 14, 'Blue');
editor.addCharacter('o', 'Times', 14, 'Blue');

const results = editor.display();
results.forEach(result => console.log(result));
